### PR TITLE
Add InstantHitWithFakeBullets

### PR DESCRIPTION
**Pull request by dnqbob**
example

Spawns several fake bullets with interval (or set interval to 0 if you want a shot gun effect) upon each fire.

### CODE REVIEW

File: OpenRA.Mods.HV/Projectiles/InstantHitWIthFakeBullets.cs [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+using OpenRA.Graphics;
+using OpenRA.Mods.Common;
+using OpenRA.Mods.Common.Effects;
+using OpenRA.Mods.Common.Graphics;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+using Util = OpenRA.Mods.Common.Util;
+
+namespace OpenRA.Mods.SP.Projectiles
```
Comment (Mailaender) [inline OpenRA.Mods.HV/Projectiles/InstantHitWIthFakeBullets.cs:24]:
```suggestion
namespace OpenRA.Mods.HV.Projectiles
```

File: OpenRA.Mods.HV/Projectiles/InstantHitWIthFakeBullets.cs [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+{
+	[Desc("Instant and usually direct-on-target projectile, with traces effect like 3rd generation of CNC.")]
+	public class InstantHitWIthFakeBulletsInfo : IProjectileInfo
```
Comment (Mailaender) [inline OpenRA.Mods.HV/Projectiles/InstantHitWIthFakeBullets.cs:27]:
```suggestion
	public class InstantHitWithFakeBulletsInfo : IProjectileInfo
```

File: OpenRA.Mods.HV/Projectiles/InstantHitWIthFakeBullets.cs [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+			public int OverallTime;
+			public ContrailRenderable Contrail;
+			public Animation Anim;
+			public WPos EndPos;
+			public WPos Pos;
+			public WPos SourcePos;
+			public WAngle Facing;
+		}
+
+		public InstantHitWIthFakeBullets(InstantHitWIthFakeBulletsInfo info, ProjectileArgs args)
```
Comment (Mailaender) [inline OpenRA.Mods.HV/Projectiles/InstantHitWIthFakeBullets.cs:149]:
```suggestion
		public InstantHitWithFakeBullets(InstantHitWIthFakeBulletsInfo info, ProjectileArgs args)
```

**Comment by OcelotWalrus** [comment]
I don't see why that'd actually be useful.

**Comment by dnqbob** [comment]
> have a weird looking shot angle

The problem is the fire muzzle sprites have initial offset which is conflicted with the "LocalOffsets" in `Armament`.
example
When set "LocalOffset: 900,0,0", the bullet trace is correct but the muzzle is wrong.

**Comment by OcelotWalrus** [comment]
I still don't find a really useful case for that new trait.

**Comment by GraionDilach** [comment]
This looks appropriate. I'd say drop the muzzle effect from the weapons using this trait though, they no longer need it (another option might be a really small circular muzzle anim akin to most RA2 gun muzzles to prevent exposing the sprite's unability to track the facing along but considering the size required for this I'd guess the size would be too small to be meaningful).

> I still don't find a really useful use for that new trait.

I can say the same about all your gameplay proposals so far.

**Comment by GraionDilach** [comment]
TBH, if you combine ammo and GrantConditionOnAttack conditions on the decoration armament, you _can_ keep the sync, but that is a massive hack (and depending on the amount of possible reloaddelaymodifier condition sources could be way too tedious/fragile), so this is indeed a more reliable option, all things considered.

Regarding the technical implementation, I'm not quite certain if FakeBulletWrapper should remain as a struct or it should be promoted to a separate IEffect subclass that InstantHitWithFakeBullets just spawns and forgets thenafter.

**Comment by OcelotWalrus** [comment]
> I can say the same about all your gameplay proposals so far.

If you think like that, then you could probably say that pretty much every pod or vehicle variant is not very useful.

**Comment by dnqbob** [comment]
> TBH, if you combine ammo and GrantConditionOnAttack conditions on the decoration armament, you can keep the sync, but that is a massive hack (and depending on the amount of possible reloaddelaymodifier condition sources could be way too tedious/fragile), so this is indeed a more reliable option, all things considered.

Yep, some painful experience, then I finally decide to make this projectile and get rid of all the hacks.

And also some mods are not developed by 1 person, and decoration weapon system is casecaded with actual weapon, which means that if the balance maker in your team modify the actual armament, it may break the behaviors of decoration armament, cause some mess.

**Comment by dnqbob** [comment]
Maybe I am a person forcus too

**Comment by dnqbob** [comment]
> I just prefer how bullets visual effects are currently.

**Comment by dnqbob** [comment]
> I just prefer how bullets visual effects are currently.

Maybe I am a person focus too much on consistency, but IMO, it is better that to make all gun weapon instant hit or all use the `Bullet` projectile that has a delay on impact. It is not very good for me to have some machine gun instant hit and some machine gun has a delay on impact with a bullet trace, which will confuse player. 

In addition, machine gun with a delay on impact behaves terrible when anti-air, which makes helicopter in HV weaker than plane in gameplay, and you say they all use "machine gun", huh.

**Comment by dnqbob** [comment]
> Regarding the technical implementation, I'm not quite certain if FakeBulletWrapper should remain as a struct or it should be promoted to a separate IEffect subclass that InstantHitWithFakeBullets just spawns and forgets thenafter.

**Comment by dnqbob** [comment]
> Regarding the technical implementation, I'm not quite certain if FakeBulletWrapper should remain as a struct or it should be promoted to a separate IEffect subclass that InstantHitWithFakeBullets just spawns and forgets thenafter.

Yes, that is not a bad idea. Then I can spawn the effect and make the parent effect of `InstantHitWithFakeBullets` end normally, not like now I have to keep the `InstantHitWithFakeBullets` when sub effect is not finished.

**Comment by GraionDilach** [comment]
Exactly. I'd say that this as-is just migrates the hackish solution from MiniYAML to C#. I also think upstream agrees with the assessment, considering that similar logic led me to design the AS particle system as-is, whose subset was merged back to upstream's FloatingSprite.

**Comment by dnqbob** [comment]
If so I am going to make an upstream version

**Comment by dnqbob** [comment]
> Regarding the technical implementation, I'm not quite certain if FakeBulletWrapper should remain as a struct or it should be promoted to a separate IEffect subclass that InstantHitWithFakeBullets just spawns and forgets thenafter.

On second thought, I find there is no other place to use the "FakeBulletTraces" effect, becuase the hit point of fake bullet depends on where the actual projectile explode, which can be blocked and even be countered. if the actual projectile cannot hit instantly, the situation will be complicated.

**Comment by GraionDilach** [comment]
Lasers (with the laser functioning as targetlaser) come to my mind as another usecase.

**Pull request merged**